src/darwin/Framework/CHIP/templates/partials/test_cluster.zapt

```diff
@@ -25,50 +30,92 @@
         {{#unless (isStrEqual "0" response.error)}}
-        [expectation fulfill];
+        {{> FulfillOrSetFinishFlag}}
         {{else}}
         {{#unless isSubscribeAttribute}}
-        {{> process_response_value}}
+
+        {{#chip_tests_item_response_parameters}}
+        {{#if hasExpectedValue}}
+        {{#if isList}}
+        XCTAssertEqual([values[@"{{#if parent.isAttribute}}value{{else}}{{name}}{{/if}}"] count], {{expectedValue.length}});
+        {{else}}
+        {{#if (isString type)}}
+        {{#if (isOctetString type)}}
+        NSString * {{asLowerCamelCase name}}ArgumentString= @"{{expectedValue}}";
+        NSData * {{asLowerCamelCase name}}Argument = [{{asLowerCamelCase name}}ArgumentString dataUsingEncoding:NSUTF8StringEncoding];
+        XCTAssertTrue([values[@"{{#if parent.isAttribute}}value{{else}}{{name}}{{/if}}"] isEqualToData:{{asLowerCamelCase name}}Argument]);
+        {{else}}
+        NSString * {{asLowerCamelCase name}}Argument= @"{{expectedValue}}";
+        XCTAssertTrue([values[@"{{#if parent.isAttribute}}value{{else}}{{name}}{{/if}}"] isEqualToString:{{asLowerCamelCase name}}Argument]);
+        {{/if}}
+        {{else}}
+        XCTAssertEqual([values[@"{{#if parent.isAttribute}}value{{else}}{{name}}{{/if}}"] {{asObjectiveCNumberType "" type true}}Value], {{expectedValue}}{{asTypeLiteralSuffix}});
+        {{/if}}
+        {{/if}}
+        {{/if}}
+        {{#if hasExpectedConstraints}}
+        {{#if expectedConstraints.minLength}}
+        XCTAssertGreaterThanOrEqual([values[@"{{#if parent.isAttribute}}value{{else}}{{name}}{{/if}}"] length], {{expectedConstraints.minLength}});
+        {{/if}}
+        {{#if expectedConstraints.maxLength}}
+        XCTAssertLessThanOrEqual([values[@"{{#if parent.isAttribute}}value{{else}}{{name}}{{/if}}"] length], {{expectedConstraints.maxLength}});
+        {{/if}}
+        {{#if expectedConstraints.minValue}}
+        XCTAssertGreaterThanOrEqual([values[@"{{#if parent.isAttribute}}value{{else}}{{name}}{{/if}}"] {{asObjectiveCNumberType "" type true}}Value], {{expectedConstraints.minValue}});
+        {{/if}}
+        {{#if expectedConstraints.maxValue}}
+        XCTAssertLessThanOrEqual([values[@"{{#if parent.isAttribute}}value{{else}}{{name}}{{/if}}"] {{asObjectiveCNumberType "" type true}}Value], {{expectedConstraints.maxValue}});
+        {{/if}}
+        {{#if expectedConstraints.notValue}}
+        XCTAssertNotEqual([values[@"{{#if parent.isAttribute}}value{{else}}{{name}}{{/if}}"] {{asObjectiveCNumberType "" type true}}Value], {{expectedConstraints.notValue}});
+        {{/if}}
+        {{/if}}
+        {{/chip_tests_item_response_parameters}}
+
         {{else}}
-        XCTAssertEqual(initialReportReceived, true);
+        XCTAssertEqual(testSendCluster{{parent.filename}}_{{asTestIndex waitForReport.index}}_{{asUpperCamelCase waitForReport.command}}_Finished, true);
         {{/unless}}
-        [expectation fulfill];
+        {{> FulfillOrSetFinishFlag}}
```

Comment:
If we remove the usage of the inline partial for the `optional`use case as well as the case where there is an error message instead of a success, the partial can probably just be removed and expanded here ?